### PR TITLE
Generalize Approximating Type Maps

### PR DESCRIPTION
**Pull request by odersky**
Approximating type maps now work with bounds instead of NoTypes. This
gives more precision (in particular for type applications), and also
makes it possible to combine approximation with other mappings.

The biggest win here is that we can now express AsSeenFrom soundly in terms of ApproximatingTypeMaps. We can therefore remove the UnsafeNonVariant annotations and 
retroactive skolemizations which felt kludgey.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 1]
```diff
@@ -3842,63 +3862,188 @@ object Types {
...
+          range(derivedAppliedType(tp, tyconLo, args), derivedAppliedType(tp, tyconHi, args))
+        case _ =>
+          if (args.exists(isRange)) {
+            if (variance > 0) tp.derivedAppliedType(tycon, args.map(rangeToBounds))
+            else {
+              val loBuf, hiBuf = new mutable.ListBuffer[Type]
+              def distributeArgs(args: List[Type], tparams: List[ParamInfo]): Boolean = args match {
+                case Range(lo, hi) :: args1 =>
+                  val v = tparams.head.paramVariance
+                  if (v == 0) false
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/core/Types.scala]: Missing "return" before "false"? Otherwise, it seems `distributeArgs` always return `true`.

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 2]
```diff
@@ -268,7 +268,20 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
...
+                      // would not detect types to be compatible. Note: If we replace the
+                      // 2nd condition by `defDenot.symbol.maybeOwner.isType` we get lots
+                      // of failures in the `tastyBootstrap` test. Trying to compile these
+                      // files in isolation works though.
+                      // TODO: Investigate why that happens.
+                      defDenot.symbol.owner
+                    else
+                      curOwner
+                  effectiveOwner.thisType.select(name, defDenot)
+                }
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala]: I still have no idea why this change becomes necessary despite the comments. Is it due to the change of variance for NamedTypes in TypeMaps?

File: compiler/src/dotty/tools/dotc/core/TypeOps.scala [hunk 3]
```diff
@@ -19,131 +19,76 @@ import ast.tpd._
...
-            asSeenFrom(tp.refinedInfo, pre, cls, theMap))
-        case tp: TypeAlias if tp.variance == 1 => // if variance != 1, need to do the variance calculation
-          tp.derivedTypeAlias(asSeenFrom(tp.alias, pre, cls, theMap))
-        case _ =>
-          (if (theMap != null) theMap else new AsSeenFromMap(pre, cls))
-            .mapOver(tp)
+          case tp: ThisType =>
+            toPrefix(pre, cls, tp.cls)
+          case _: BoundType | NoPrefix =>
+            tp
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/core/TypeOps.scala]: I believe `tp == NoType` will only happen in exceptional cases. So we don't want to slow down the general logic by testing for it early.

Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/core/TypeOps.scala]: Actually during the debugging, it's quite common, triggered from the Namer: https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/typer/Namer.scala#L1011

Comment (odersky) [inline compiler/src/dotty/tools/dotc/core/TypeOps.scala]: That's probably still a lot less than the other cases. To be sure either way we'd have to measure (i.e. add counters).

Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/core/TypeOps.scala]: I agree, it's better to handle that special case in `Namer` instead of `AsSeenFromMap`.

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 4]
```diff
@@ -3842,63 +3862,188 @@ object Types {
...
   }
 
-  /** A type map that approximates NoTypes by upper or lower known bounds depending on
+  case class Range(lo: Type, hi: Type) extends UncachedGroundType {
+    assert(!lo.isInstanceOf[Range])
+    assert(!hi.isInstanceOf[Range])
+
+    override def toText(printer: Printer): Text =
+      lo.toText(printer) ~ ".." ~ hi.toText(printer)
+  }
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/core/Types.scala]: Yes, good idea.

File: compiler/src/dotty/tools/dotc/typer/Typer.scala (hunk 2 above)
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala]: I am not sure.

File: compiler/src/dotty/tools/dotc/printing/PlainPrinter.scala [hunk 5]
```diff
@@ -68,6 +68,15 @@ class PlainPrinter(_ctx: Context) extends Printer {
       }
     else tp
 
+  private def sameBound(lo: Type, hi: Type): Boolean =
+    try lo =:= hi
+    catch { case ex: Throwable => false }
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/printing/PlainPrinter.scala]: I think this is a bit sloppy :). I would like `NonFatal(ex)` more, or even better: define a base trait that all non-fatal exceptions we throw in the typechecker must extend. (Or even even better: make the high-level APIs of the typer deal with exceptions themselves, clearly there is no value in propagating the exceptions up more if we're just ignoring them).

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 6]
```diff
@@ -3851,63 +3860,193 @@ object Types {
...
+          range(derivedRefinedType(tp, parentLo, info), derivedRefinedType(tp, parentHi, info))
+        case _ =>
+          if (parent.isBottomType) parent
+          else info match {
+            case Range(infoLo, infoHi) =>
+              def propagate(lo: Type, hi: Type) =
+                range(derivedRefinedType(tp, parent, lo), derivedRefinedType(tp, parent, hi))
+              tp.refinedInfo match {
+                case rinfo: TypeBounds =>
+                  val v = if (rinfo.isAlias) rinfo.variance * variance else variance
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/Types.scala]: I don't understand what's going on here, could you add some comments? I've also checked and we have zero test cases where v does not end up being 0 here, so it's hard to judge the validity of the code.

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 7]
```diff
@@ -3851,63 +3860,193 @@ object Types {
...
+                  if (v > 0) tp.derivedRefinedType(parent, tp.refinedName, rangeToBounds(info))
+                  else if (v < 0) propagate(infoHi, infoLo)
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/Types.scala]: I find this especially suspicious, this seems to violate the invariant that `Range#lo` is a subtype of `Range#hi`, am I missing something?

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 8]
```diff
@@ -3851,63 +3860,193 @@ object Types {
...
-      if (tycon.exists && args.forall(_.exists)) tp.derivedAppliedType(tycon, args)
-      else approx() // This is rather coarse, but to do better is a bit complicated
+      tycon match {
+        case Range(tyconLo, tyconHi) =>
+          range(derivedAppliedType(tp, tyconLo, args), derivedAppliedType(tp, tyconHi, args))
+        case _ =>
+          if (args.exists(isRange)) {
+            if (variance > 0) tp.derivedAppliedType(tycon, args.map(rangeToBounds))
+            else {
+              val loBuf, hiBuf = new mutable.ListBuffer[Type]
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/Types.scala]: I think this case is also complicated enough to warrant a comment :).

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 9]
```diff
@@ -3851,63 +3860,193 @@ object Types {
...
+      case tp: Range => tp.hi
+      case _ => tp
+    }
+
+    protected def rangeToBounds(tp: Type) = tp match {
+      case Range(lo, hi) => TypeBounds(lo, hi)
+      case _ => tp
+    }
+
+    protected def atVariance[T](v: Int)(op: => T): T = {
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/Types.scala]: I think this is worth putting in `TypeMap` itself, considering how common the pattern `val saved = variance; variance = ...` is. Also maybe worth marking `inline` ?

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 10]
```diff
@@ -3851,63 +3860,193 @@ object Types {
...
    *     variance < 0 : approximate by lower bound
-   *     variance = 0 : propagate NoType to next outer level
+   *     variance = 0 : propagate bounds to next outer level
    */
   abstract class ApproximatingTypeMap(implicit ctx: Context) extends TypeMap { thisMap =>
-    def approx(lo: Type = defn.NothingType, hi: Type = defn.AnyType) =
-      if (variance == 0) NoType
-      else apply(if (variance < 0) lo else hi)
 
+    protected def range(lo: Type, hi: Type) =
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/Types.scala]: Calling this method `range` makes it a bit too easy to confuse it with `Range`. Perhaps call it `between` ?

Comment (odersky) [inline compiler/src/dotty/tools/dotc/core/Types.scala]: The current situation is analogous to `range` / `Range` for collection ranges. I believe that's a good thing.

File: compiler/src/dotty/tools/dotc/printing/PlainPrinter.scala [hunk 11]
```diff
@@ -68,6 +68,15 @@ class PlainPrinter(_ctx: Context) extends Printer {
...
+
+  private def homogenizeArg(tp: Type) = tp match {
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/printing/PlainPrinter.scala]: In fact, the name was correct, but the condition to do this only under `homogenizedView` was lacking.

**Comment by odersky** [comment]
I'll roll the comment fixes into #2945. Can you review that one next? 😄

**Comment by smarter** [comment]
Sure!

**Comment by liufengyun** [comment]
test performance please

**Comment by smarter** [comment]
Merged as part of https://github.com/lampepfl/dotty/pull/2970 (and re-merged as part of https://github.com/lampepfl/dotty/pull/2945)
